File: services/gateway/src/services/posts/__tests__/reelComposition.test.ts
/**
 * @jest-environment node
 *
 * Règle de composition d'un RÉEL (directive user 2026-08-02, étendue par la
 * directive durée minimale) : video (>=3s) || audio (>=3s) || >= 2 images —
 * miroir exact du SDK (`ReelComposition.qualifiesAsReel`, FeedModels.swift,
 * suite `ReelCompositionTests`). Les cas ci-dessous répliquent la suite Swift.
 * Les images ne sont jamais soumises à la condition de durée.
 *
 * `qualifiesAsReel` vit désormais dans `@meeshy/shared/utils/reel-composition`
 * (source unique, testée aussi côté shared dans
 * `packages/shared/__tests__/utils/reel-composition.test.ts`) — ce fichier
 * reste une couverture de régression au point de consommation gateway.
 */

import { describe, it, expect } from '@jest/globals';
import { qualifiesAsReel } from '@meeshy/shared/utils/reel-composition';

const LONG = 5000;

const media = (...entries: Array<string | null | [string | null, number | null | undefined]>) =>
  entries.map((entry) =>
    Array.isArray(entry)
      ? { mimeType: entry[0], duration: entry[1] }
      : { mimeType: entry, duration: LONG },
  );

describe('qualifiesAsReel', () => {
  it('qualifies a video, an audio, or at least two images (sufficient duration)', () => {
    expect(qualifiesAsReel(media('video/mp4'))).toBe(true);
    expect(qualifiesAsReel(media('audio/mp4'))).toBe(true);
    expect(qualifiesAsReel(media('image/jpeg', 'image/png'))).toBe(true);
    expect(qualifiesAsReel(media('image/jpeg', 'image/png', 'image/heic'))).toBe(true);
    expect(qualifiesAsReel(media('audio/mpeg', 'image/heic'))).toBe(true);
    expect(qualifiesAsReel(media('video/quicktime', 'image/jpeg'))).toBe(true);
  });

  it('does NOT qualify a single image — the 2→1 removal trap', () => {
    expect(qualifiesAsReel(media('image/jpeg'))).toBe(false);
    expect(qualifiesAsReel(media('image/jpeg', 'application/pdf'))).toBe(false);
  });

  it('never qualifies an empty or non-media composition', () => {
    expect(qualifiesAsReel([])).toBe(false);
    expect(qualifiesAsReel(media('application/pdf'))).toBe(false);
    expect(qualifiesAsReel(media('application/pdf', 'text/plain'))).toBe(false);
  });

  it('treats a null or unknown mimeType as non-qualifying', () => {
    expect(qualifiesAsReel(media(null))).toBe(false);
    expect(qualifiesAsReel(media(null, null))).toBe(false);
    expect(qualifiesAsReel(media('video/mp4', null))).toBe(true);
  });

  it('is case-insensitive on the MIME type', () => {
    expect(qualifiesAsReel(media('IMAGE/JPEG', 'Image/PNG'))).toBe(true);
    expect(qualifiesAsReel(media('VIDEO/MP4'))).toBe(true);
    expect(qualifiesAsReel(media('IMAGE/JPEG'))).toBe(false);
  });

  it('does NOT qualify a video or audio under 3 seconds', () => {
    expect(qualifiesAsReel(media(['video/mp4', 2999]))).toBe(false);
    expect(qualifiesAsReel(media(['audio/mp4', 2999]))).toBe(false);
    expect(qualifiesAsReel(media(['video/mp4', 0]))).toBe(false);
  });

  it('qualifies a video or audio at exactly 3 seconds', () => {
    expect(qualifiesAsReel(media(['video/mp4', 3000]))).toBe(true);
    expect(qualifiesAsReel(media(['audio/mp4', 3000]))).toBe(true);
  });

  it('treats a missing or null duration on video/audio as non-qualifying', () => {
    expect(qualifiesAsReel(media(['video/mp4', null]))).toBe(false);
    expect(qualifiesAsReel(media(['video/mp4', undefined]))).toBe(false);
    expect(qualifiesAsReel(media(['audio/mp4', null]))).toBe(false);
  });

  it('never applies the duration floor to images', () => {
    expect(qualifiesAsReel(media(['image/jpeg', 0], ['image/png', null]))).toBe(true);
  });

  it('a short video is rescued by a second qualifying image pair', () => {
    expect(qualifiesAsReel(media(['video/mp4', 500], 'image/jpeg', 'image/png'))).toBe(true);
  });
});

File: apps/android/core/model/src/test/kotlin/me/meeshy/sdk/model/ReelCompositionTest.kt
package me.meeshy.sdk.model

import com.google.common.truth.Truth.assertThat
import org.junit.Test

/**
 * Behavioural spec for [ReelComposition] — the creation-time reel-qualification
 * predicate (règle produit 2026-08-02 + directive durée minimale). Mirrors iOS SDK
 * `ReelCompositionTests` (packages/MeeshySDK/Tests/MeeshySDKTests/Models/
 * FeedReelClassificationTests.swift) and the gateway's authoritative
 * `qualifiesAsReel` (services/gateway/src/services/posts/reelComposition.ts) —
 * every case here has a same-shaped sibling on both other sites.
 */
class ReelCompositionTest {

    /** Duration comfortably above the floor — used by cases unrelated to timing. */
    private val long = 5000L

    private fun media(vararg mimeTypes: String): List<ReelComposition.MediaKind> =
        mimeTypes.map { ReelComposition.MediaKind(mimeType = it, durationMs = long) }

    // --- qualifiesAsReel — video (>=3s) || audio (>=3s) || >= 2 images ------

    @Test
    fun `a video, an audio, or at least two images qualifies as a reel`() {
        assertThat(ReelComposition.qualifiesAsReel(media("video/mp4"))).isTrue()
        assertThat(ReelComposition.qualifiesAsReel(media("audio/mp4"))).isTrue()
        assertThat(ReelComposition.qualifiesAsReel(media("image/jpeg", "image/jpeg"))).isTrue()
        assertThat(ReelComposition.qualifiesAsReel(media("image/jpeg", "image/jpeg", "image/jpeg"))).isTrue()
        assertThat(ReelComposition.qualifiesAsReel(media("audio/mp4", "image/jpeg"))).isTrue()
        assertThat(ReelComposition.qualifiesAsReel(media("video/mp4", "image/jpeg"))).isTrue()
    }

    @Test
    fun `a single image does NOT qualify — the 2 to 1 removal trap`() {
        assertThat(ReelComposition.qualifiesAsReel(media("image/jpeg"))).isFalse()
        assertThat(ReelComposition.qualifiesAsReel(media("image/jpeg", "application/pdf"))).isFalse()
    }

    @Test
    fun `no media or only non-reel media never qualifies`() {
        assertThat(ReelComposition.qualifiesAsReel(emptyList())).isFalse()
        assertThat(ReelComposition.qualifiesAsReel(media("application/pdf"))).isFalse()
        assertThat(ReelComposition.qualifiesAsReel(media("application/pdf", "application/pdf"))).isFalse()
    }

    @Test
    fun `a video or audio under 3 seconds does NOT qualify`() {
        assertThat(
            ReelComposition.qualifiesAsReel(listOf(ReelComposition.MediaKind("video/mp4", 2999))),
        ).isFalse()
        assertThat(
            ReelComposition.qualifiesAsReel(listOf(ReelComposition.MediaKind("audio/mp4", 2999))),
        ).isFalse()
        assertThat(
            ReelComposition.qualifiesAsReel(listOf(ReelComposition.MediaKind("video/mp4", 0))),
        ).isFalse()
    }

    @Test
    fun `a video or audio at exactly 3 seconds qualifies`() {
        assertThat(
            ReelComposition.qualifiesAsReel(listOf(ReelComposition.MediaKind("video/mp4", 3000))),
        ).isTrue()
        assertThat(
            ReelComposition.qualifiesAsReel(listOf(ReelComposition.MediaKind("audio/mp4", 3000))),
        ).isTrue()
    }

    @Test
    fun `a missing duration on video or audio does NOT qualify`() {
        assertThat(
            ReelComposition.qualifiesAsReel(listOf(ReelComposition.MediaKind("video/mp4", null))),
        ).isFalse()
        assertThat(
            ReelComposition.qualifiesAsReel(listOf(ReelComposition.MediaKind("audio/mp4", null))),
        ).isFalse()
    }

    @Test
    fun `images are never subject to the duration floor`() {
        assertThat(
            ReelComposition.qualifiesAsReel(
                listOf(ReelComposition.MediaKind("image/jpeg", 0), ReelComposition.MediaKind("image/png", null)),
            ),
        ).isTrue()
    }

    @Test
    fun `mime matching is case-insensitive`() {
        assertThat(ReelComposition.qualifiesAsReel(media("IMAGE/JPEG", "Image/PNG"))).isTrue()
    }

    // --- defaultType ---------------------------------------------------------

    @Test
    fun `qualifying compositions default to REEL`() {
        assertThat(ReelComposition.defaultType(media("video/mp4"))).isEqualTo(PostType.REEL)
        assertThat(ReelComposition.defaultType(media("audio/mp4"))).isEqualTo(PostType.REEL)
        assertThat(ReelComposition.defaultType(media("image/jpeg", "image/jpeg"))).isEqualTo(PostType.REEL)
    }

    @Test
    fun `a single image defaults to POST even without forcePlainPost`() {
        assertThat(ReelComposition.defaultType(media("image/jpeg"))).isEqualTo(PostType.POST)
    }

    @Test
    fun `a video under 3 seconds defaults to POST`() {
        assertThat(
            ReelComposition.defaultType(listOf(ReelComposition.MediaKind("video/mp4", 1000))),
        ).isEqualTo(PostType.POST)
    }

    @Test
    fun `forcing a plain post overrides the reel default`() {
        assertThat(ReelComposition.defaultType(media("video/mp4"), forcePlainPost = true)).isEqualTo(PostType.POST)
        assertThat(ReelComposition.defaultType(media("audio/mp4"), forcePlainPost = true)).isEqualTo(PostType.POST)
    }

    @Test
    fun `text-only or document-only compositions default to POST`() {
        assertThat(ReelComposition.defaultType(emptyList())).isEqualTo(PostType.POST)
        assertThat(ReelComposition.defaultType(media("application/pdf"))).isEqualTo(PostType.POST)
    }
}
